drivers/power/adp1050/adp1050.c

```diff
@@ -0,0 +1,1250 @@
+				     new_pass, 1);
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * @brief Software reset the ADP1050.
+ * @param desc - ADP1050 device descriptor
+ * @return 0 in case of succes, negative error code otherwise
+*/
+int adp1050_software_reset(struct adp1050_desc *desc)
+{
+	int ret;
+
+	ret = adp1050_write(desc, ADP1050_SOFTWARE_RESET_SETTINGS,
+			    ADP1050_SW_RES_DELAY_500MS
+			    | ADP1050_SW_RES_NO_DELAY, 1);
+	if (ret)
+		return ret;
+
+	return adp1050_write(desc, ADP1050_SOFTWARE_RESET_GO,
+			     ADP1050_SOFTWARE_RESET_ON, 1);
+}
+
+/**
+ * @brief Initialize the ADP1050 device.
+ * @param desc - ADP1050 device descriptor
+ * @param init_param - Initialization parameter containing information about the
+ * 		       ADP1050 device to be initialized.
+ * @return 0 in case of succes, negative error code otherwise
+*/
+int adp1050_init(struct adp1050_desc **desc,
+		 struct adp1050_init_param *init_param)
+{
+	struct adp1050_desc *descriptor;
+	int ret;
+
+	descriptor = (struct adp1050_desc *)no_os_calloc(sizeof(*descriptor), 1);
+	if (!descriptor)
+		return -ENOMEM;
+
+	ret = no_os_i2c_init(&descriptor->i2c_desc, init_param->i2c_param);
+	if (ret)
+		goto free_desc;
+
+	ret = no_os_gpio_get_optional(&descriptor->pg_alt_desc,
+				      init_param->pg_alt_param);
+	if (ret)
+		goto free_i2c;
+
+	ret = no_os_gpio_direction_input(descriptor->pg_alt_desc);
+	if (ret)
+		goto free_pgalt;
+
+	if (init_param->ext_syni) {
+		ret = no_os_pwm_init(&descriptor->syni_desc,
+				     init_param->syni_param);
+		if (ret)
+			goto free_pgalt;
+
+		descriptor->flgi_desc = NULL;
+	} else {
+		ret = no_os_gpio_get_optional(&descriptor->flgi_desc,
+					      init_param->flgi_param);
+		if (ret)
+			goto free_pgalt;
+
+		ret = no_os_gpio_direction_output(descriptor->flgi_desc,
+						  NO_OS_GPIO_LOW);
+		if (ret)
+			goto free_flgi;
+
+		descriptor->syni_desc = NULL;
+	}
+
+	ret = no_os_gpio_get_optional(&descriptor->ctrl_desc,
+				      init_param->ctrl_param);
+	if (ret)
+		goto free_flgi;
+
+	if (descriptor->ctrl_desc
+	    && !no_os_field_get(ADP1050_CTRL_PIN_ENABLE, init_param->on_off_config))
+		goto free_ctrl;
+
+	ret = adp1050_write(descriptor, ADP1050_ON_OFF_CONFIG,
+			    (uint16_t)init_param->on_off_config, 1);
+	if (ret)
+		goto free_ctrl;
+
+	ret = adp1050_write(descriptor, ADP1050_OPERATION, ADP1050_OPERATION_ON, 1);
+	if (ret)
+		goto free_ctrl;
+
+	/* Time needed for the ADP1050 to power-up. */
+	no_os_mdelay(52);
+
+	ret = adp1050_send_command(descriptor, ADP1050_CLEAR_FAULTS);
+	if (ret)
+		goto free_ctrl;
+
+	descriptor->loop = ADP1050_CLOSE_LOOP;
+	descriptor->freq = ADP1050_49KHZ;
+
+	*desc = descriptor;
+
+	return 0;
+
+free_ctrl:
+	no_os_gpio_remove(descriptor->ctrl_desc);
+free_flgi:
+	no_os_gpio_remove(descriptor->flgi_desc);
+	no_os_pwm_remove(descriptor->syni_desc);
+free_pgalt:
+	no_os_gpio_remove(descriptor->pg_alt_desc);
+free_i2c:
+	no_os_i2c_remove(descriptor->i2c_desc);
+free_desc:
+	no_os_free(descriptor);
```

Comment:
I would remove all this labels, and leave only one in which you call adp1050_remove

Reply:
Will do this and add an if condition in the remove function to see if the i2c_desc is initialized, otherwise the remove function will return -EINVAL, without freeing the memory.